test/es6/es6HasInstance.js

```diff
@@ -3,86 +3,396 @@
         }
-    });
-
-    var x = new proxy();
-    print((x) instanceof proxy);
-}
-
-// proxy of bounded function
-function test4() {
-    print('*** test4');
-    function foo() { };
-    Object.defineProperty(foo, Symbol.hasInstance, {
-        value: function () {
-            print('checked Symbol.hasInstance');
-            return false;
+    },
+    {
+        name: "instanceof operator invokes [[getPrototypeOf]] internal method on left-hand side value",
+        body: function() {
+
+            var p = new Proxy({}, {
+                getPrototypeOf: function() {
+                    throw new Error('Hit getPrototypeOf');
+                }
+            });
+            var obj = Object.create(p);
+            obj.prototype = {};
+            var F = function() {};
+
+            assert.throws(()=>{p instanceof F}, Error, "p instanceof F", 'Hit getPrototypeOf');
+            assert.throws(()=>{obj instanceof F}, Error, "obj instanceof F", 'Hit getPrototypeOf');
         }
-    });
-    var proxy = new Proxy(foo, {
-        get: function (target, property) {
-            print(property.toString());
-            return Reflect.get(target, property);
+    },
+    {
+        name: "changing Symbol.hasIstance property on a function contructor invalidates inline cache",
+        body: function() {
+            var F = function() {}
+
+            var changeHasInstance = function() {
+                Object.defineProperty(F, Symbol.hasInstance, {
+                    value: function(inst) { return true; }
+                });
+            }
+
+            function func() {
+                return 0 instanceof F;
+            }
+
+            var changed = false;
+            for (var i=0; i<100; i++) {
+                var x = func();
+                assert.areEqual(changed, x, "i=="+i);
+                if (i==20) {
+                    changeHasInstance();
+                    changed = true;
+                }
+            }
+
+        }
+    },
+    {
+        name: "changing 'prototype' property on a function contructor invalidates inline cache",
+        body: function() {
+            var F = function() {}
+
+            var changeHasInstance = function() {
+                Object.defineProperty(F, 'prototype', {
+                    value: Function.prototype
+                });
+            }
+
+            function func() {
+                return function(){} instanceof F;
+            }
+
+            var changed = false;
+            for (var i=0; i<100; i++) {
+                var x = func();
+                assert.areEqual(changed, x, "i=="+i);
+                if (i==20) {
+                    changeHasInstance();
+                    changed = true;
+                }
+            }
+
         }
-    });
+    },
+];
 
-    var x = proxy.bind();
-    print((x) instanceof proxy);
-}
+testRunner.runTests(tests, { verbose: WScript.Arguments[0] != "summary" });

```

Comment:
Confirmed no test262 or perf regressions.

---
In reply to: [65278274](https://github.com/Microsoft/ChakraCore/pull/1063#discussion_r65278274) [](ancestors = 65278274)